# Weekly Assignment 05

In [260]:
import pandas as pd
import sklearn as sk
import matplotlib.pyplot as plt

In [261]:
df = pd.read_csv("student-por.csv")
df.head()

school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  ...  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher  ...   
1     GP   F   17       U     GT3       T     1     1  at_home     other  ...   
2     GP   F   15       U     LE3       T     1     1  at_home     other  ...   
3     GP   F   15       U     GT3       T     4     2   health  services  ...   
4     GP   F   16       U     GT3       T     3     3    other     other  ...   

  famrel freetime  goout  Dalc  Walc health absences  G1  G2  G3  
0      4        3      4     1     1      3        4   0  11  11  
1      5        3      3     1     1      3        2   9  11  11  
2      4        3      2     2     3      3        6  12  13  12  
3      3        2      2     1     1      5        0  14  14  14  
4      4        3      2     1     2      5        0  11  13  13  

[5 rows x 33 columns]

In [262]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 649 entries, 0 to 648
Data columns (total 33 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   school      649 non-null    object
 1   sex         649 non-null    object
 2   age         649 non-null    int64 
 3   address     649 non-null    object
 4   famsize     649 non-null    object
 5   Pstatus     649 non-null    object
 6   Medu        649 non-null    int64 
 7   Fedu        649 non-null    int64 
 8   Mjob        649 non-null    object
 9   Fjob        649 non-null    object
 10  reason      649 non-null    object
 11  guardian    649 non-null    object
 12  traveltime  649 non-null    int64 
 13  studytime   649 non-null    int64 
 14  failures    649 non-null    int64 
 15  schoolsup   649 non-null    object
 16  famsup      649 non-null    object
 17  paid        649 non-null    object
 18  activities  649 non-null    object
 19  nursery     649 non-null    object
 20  higher    

# Data Processing 

For our hypothesis, we are going to select the below variables in our model and see if we can get an accurate prediction:
 
* school
* sex
* address: 
* Parents cohabitation
* guardian
* studytime
* internet
* famrel
* freetime
* goout
* absence
* G1
* G2
* G3 

In [263]:
# subseting the original Dataframe
df_subset = df[["school", "sex", "address", "Pstatus","guardian", "studytime", "internet", "famrel", "freetime", "goout", "absences", "G1", "G2", "G3", "romantic"]]
df
df_subset.head()

school sex address Pstatus guardian  studytime internet  famrel  freetime  \
0     GP   F       U       A   mother          2       no       4         3   
1     GP   F       U       T   father          2      yes       5         3   
2     GP   F       U       T   mother          2      yes       4         3   
3     GP   F       U       T   mother          3      yes       3         2   
4     GP   F       U       T   father          2       no       4         3   

   goout  absences  G1  G2  G3 romantic  
0      4         4   0  11  11       no  
1      3         2   9  11  11       no  
2      2         6  12  13  12       no  
3      2         0  14  14  14      yes  
4      2         0  11  13  13       no

We choose categorical variables and turn them into dummy variables. The way we do so is by creating a list and loop around it to change the names of the new column and place them in a single Dataframe.

In [264]:
list_to_dummies = df_subset[["school", "sex", "address", "Pstatus", "guardian", "studytime", "internet", 
                            "famrel", "goout", "romantic"]].columns.tolist()

list_to_dummies

['school',
 'sex',
 'address',
 'Pstatus',
 'guardian',
 'studytime',
 'internet',
 'famrel',
 'goout',
 'romantic']

In [265]:
# looping through the list and naming the variables for proper distinction
df_temp = pd.DataFrame()
for item in list_to_dummies:
    dummy = pd.get_dummies(df_subset[item])
    dummy_cols = dummy.columns.tolist()
    for cols in dummy_cols:
        dummy = dummy.rename(columns={cols : str(item)+"_"+str(cols)})
    df_temp = pd.concat([df_temp,dummy], axis = 1)
df_temp.head()

school_GP  school_MS  sex_F  sex_M  address_R  address_U  Pstatus_A  \
0          1          0      1      0          0          1          1   
1          1          0      1      0          0          1          0   
2          1          0      1      0          0          1          0   
3          1          0      1      0          0          1          0   
4          1          0      1      0          0          1          0   

   Pstatus_T  guardian_father  guardian_mother  ...  famrel_3  famrel_4  \
0          0                0                1  ...         0         1   
1          1                1                0  ...         0         0   
2          1                0                1  ...         0         1   
3          1                0                1  ...         1         0   
4          1                1                0  ...         0         1   

   famrel_5  goout_1  goout_2  goout_3  goout_4  goout_5  romantic_no  \
0         0        0        0        0        1        0            1   
1         1        0        0        1        0        0            1   
2         0        0        1        0        0        0            1   
3         0        0        1        0        0        0            0   
4         0        0        1        0        0        0            1   

   romantic_yes  
0             0  
1             0  
2             0  
3             1  
4             0  

[5 rows x 29 columns]

In [266]:
#adding quantitative variables to our new dataframe 
df_temp = pd.concat([df_temp, df_subset.loc[:,"G1":"G3"]], axis = 1)
df_temp = pd.concat([df_temp, df_subset["romantic"]], axis = 1)
df_temp

school_GP  school_MS  sex_F  sex_M  address_R  address_U  Pstatus_A  \
0            1          0      1      0          0          1          1   
1            1          0      1      0          0          1          0   
2            1          0      1      0          0          1          0   
3            1          0      1      0          0          1          0   
4            1          0      1      0          0          1          0   
..         ...        ...    ...    ...        ...        ...        ...   
644          0          1      1      0          1          0          0   
645          0          1      1      0          0          1          0   
646          0          1      1      0          0          1          0   
647          0          1      0      1          0          1          0   
648          0          1      0      1          1          0          0   

     Pstatus_T  guardian_father  guardian_mother  ...  goout_2  goout_3  \
0            0                0                1  ...        0        0   
1            1                1                0  ...        0        1   
2            1                0                1  ...        1        0   
3            1                0                1  ...        1        0   
4            1                1                0  ...        1        0   
..         ...              ...              ...  ...      ...      ...   
644          1                0                1  ...        1        0   
645          1                0                1  ...        0        0   
646          1                0                1  ...        0        0   
647          1                0                1  ...        0        0   
648          1                0                1  ...        0        0   

     goout_4  goout_5  romantic_no  romantic_yes  G1  G2  G3  romantic  
0          1        0            1             0   0  11  11        no  
1          0        0            1             0   9  11  11        no  
2          0        0            1             0  12  13  12        no  
3          0        0            0             1  14  14  14       yes  
4          0        0            1             0  11  13  13        no  
..       ...      ...          ...           ...  ..  ..  ..       ...  
644        0        0            1             0  10  11  10        no  
645        1        0            1             0  15  15  16        no  
646        0        0            1             0  11  12   9        no  
647        0        1            1             0  10  10  10        no  
648        0        0            1             0  10  11  11        no  

[649 rows x 33 columns]

In [267]:
# checking the value count of romantic variable
df_temp["romantic"].value_counts()

no     410
yes    239
Name: romantic, dtype: int64

In [268]:
# changing the variable from yes and no to 0 and 1 for later use
def change_yesno_to_numbers(x):
    if x == "yes":
        return("1")
    else:
        return("0")

df_temp["romantic"] = df_temp["romantic"].apply(change_yesno_to_numbers)
df_temp["romantic"] = df_temp["romantic"].astype(int)
df_temp["romantic"]

0      0
1      0
2      0
3      1
4      0
      ..
644    0
645    0
646    0
647    0
648    0
Name: romantic, Length: 649, dtype: int64

In [269]:
df_temp.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 649 entries, 0 to 648
Data columns (total 33 columns):
 #   Column           Non-Null Count  Dtype
---  ------           --------------  -----
 0   school_GP        649 non-null    uint8
 1   school_MS        649 non-null    uint8
 2   sex_F            649 non-null    uint8
 3   sex_M            649 non-null    uint8
 4   address_R        649 non-null    uint8
 5   address_U        649 non-null    uint8
 6   Pstatus_A        649 non-null    uint8
 7   Pstatus_T        649 non-null    uint8
 8   guardian_father  649 non-null    uint8
 9   guardian_mother  649 non-null    uint8
 10  guardian_other   649 non-null    uint8
 11  studytime_1      649 non-null    uint8
 12  studytime_2      649 non-null    uint8
 13  studytime_3      649 non-null    uint8
 14  studytime_4      649 non-null    uint8
 15  internet_no      649 non-null    uint8
 16  internet_yes     649 non-null    uint8
 17  famrel_1         649 non-null    uint8
 18  famrel_2  

In order to avoid colinearity and dummy variable trap, we select _n - 1_ where _n_ is the number of dummy variable for every category

In [270]:
# instead of selecting each variable, we put all the column names in a list and removing the ones we dont need
list_of_X = df_temp.columns.tolist()
list_of_X.remove("school_MS")
list_of_X.remove("sex_M")
list_of_X.remove("address_U")
list_of_X.remove("Pstatus_T")
list_of_X.remove("guardian_other")
list_of_X.remove("studytime_4")
list_of_X.remove("internet_yes")
list_of_X.remove("famrel_5")
list_of_X.remove("goout_5")
list_of_X.remove("romantic_yes")
list_of_X.remove("romantic_no")
list_of_X.remove("romantic")
X = df_temp[list_of_X]
y = df_temp["romantic"]

# Knn Model

for this model we are going to use K Nearest Neighbor algorithm

Knn algorithm identifies new observations based on their  Eucledian distance from their nearest neighbors. In our case we are going to choose _k = 3_

In [271]:
# spliting the model into training set and test set
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=1)

In [272]:
# creating an instance of the knn model from sklearn library
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier()

In [273]:
# training the model
knn = KNeighborsClassifier(n_neighbors=3)
knn = knn.fit(X_train, y_train)


In [274]:
knn.score(X_test, y_test)

0.6256410256410256

our Knn score is **0.625**, which means **62%** of our model prediction is accurate

In [275]:
# creating a confusion matrix 
from sklearn.metrics import confusion_matrix
y_pred = knn.predict(X_test) 
cm = confusion_matrix(y_test, y_pred) 
cm = pd.DataFrame(cm, index=['No (actual)', 'Yes (actual)'], columns = ['No (pred)', 'Yes (pred)']) #label and make df
cm

No (pred)  Yes (pred)
No (actual)         104          24
Yes (actual)         49          18

Now we are going to predict the accuracy of our model

In [253]:
(104+18)/(104+24+49+18)

0.6256410256410256

The calcualtion verifies our previous results. **62.5%** of our predictions are accurate.

**Precision (No Romantic Relationship)**

We are going to check the precision of no romantic relationship in our model.

Correct Predicted No Romantic Relationship / Total Number of No Romantic RelationshipPrediction

In [277]:
104/(104+49)

0.6797385620915033

We get a value of almost **68%**, which means out of the total number of predicted **No Romantic Relationship**, 68% are predicted correctly

**Recall**

In [279]:
(104)/(104+24)

0.8125

Our Recall is **81%**, which means out of the 104 predicted No Romantic Relationship, 81% are accurate predictions